File: src/python/lateral_pru.py
"""LateralPRU - lateral traverse axis wrapper around IpcChannel.

Mirrors LateralPRU.cpp / LateralPRU.h.
Klipper-style: all steps pushed as (interval, count, add) move triples via sendMove().
CMD_HOME_START triggers PRU1-side homing.  No CMD_LATERAL_* commands.
"""
from ipc import (
    IpcChannel,
    CMD_ENABLE, CMD_EMERGENCY_STOP, CMD_RESET_POSITION,
    CMD_QUEUE_FLUSH, CMD_HOME_START,
    AXIS_LATERAL, AXIS_ALL,
    STATE_RUNNING, STATE_AT_HOME,
    FAULT_HOME_SENSOR, FAULT_OVERRUN,
)
from config import (
    LAT_STEPS_PER_MM, LAT_HOME_SPEED_HZ, LAT_HOME_OFFSET_DEFAULT_MM,
    STEPS_PER_REV as SPINDLE_STEPS_PER_REV,
)

# State strings
FAULT          = "FAULT"
HOMING         = "HOMING"
HOMING_OFFSET  = "HOMING_OFFSET"
HOMED          = "HOMED"
POSITIONING    = "POSITIONING"
WINDING_FWD    = "WINDING_FWD"
WINDING_BWD    = "WINDING_BWD"

PRU_CLOCK_HZ = 200_000_000
MIN_INTERVAL  = 625
_LAT_MIN_HZ   = 100
_REFILL_MS    = 20


def _hz_to_interval(hz: int) -> int:
    if hz <= 0:
        return 0
    return max(MIN_INTERVAL, PRU_CLOCK_HZ // (2 * hz))


def _mm_to_steps(mm: float) -> int:
    return max(0, int(mm * LAT_STEPS_PER_MM)) if LAT_STEPS_PER_MM else 0


def _steps_to_mm(steps: int) -> float:
    return steps / LAT_STEPS_PER_MM if LAT_STEPS_PER_MM else 0.0


def _push_move_mm(ch: IpcChannel, hz: int, forward: bool, dist_mm: float) -> bool:
    """Push a constant-speed block covering dist_mm at hz."""
    if hz <= 0 or dist_mm < 0.001:
        return False
    n_steps = _mm_to_steps(dist_mm)
    if n_steps <= 0:
        return False
    iv = _hz_to_interval(hz)
    direction = 0 if forward else 1
    return ch.sendMove(AXIS_LATERAL, iv, 2 * n_steps, 0, direction)


def _push_constant_ms(ch: IpcChannel, hz: int, forward: bool, duration_ms: int) -> bool:
    """Push a constant-speed block for approximately duration_ms."""
    if hz <= 0 or duration_ms <= 0:
        return False
    iv = _hz_to_interval(hz)
    edges = max(1, (2 * hz * duration_ms) // 1000)
    direction = 0 if forward else 1
    return ch.sendMove(AXIS_LATERAL, iv, edges, 0, direction)


class LateralPRU:
    """Lateral axis - Linux-side state machine delegating to PRU1 via IpcChannel."""

    def __init__(self, channel: IpcChannel):
        self._ch = channel
        self._home_offset_mm: float = float(LAT_HOME_OFFSET_DEFAULT_MM)
        self._state: str = HOMING
        self._telem: dict = {}

        # Motion targets
        self._start_pos_mm: float = 0.0
        self._start_mm: float = 0.0
        self._end_mm:   float = 0.0

        # Pass tracking
        self._pass_count: int = 0
        self._last_lat_hz: int = 0
        self._winding_fwd: bool = True

        # One-shot flags
        self._stop_on_next_high:     bool = False
        self._stop_on_next_low:      bool = False
        self._pause_on_next_reversal: bool = False
        self._paused_at_reversal:    bool = False

    # -- Init -----------------------------------------------------------
    def begin(self, home_offset_mm: float = None):
        if home_offset_mm is not None:
            self._home_offset_mm = home_offset_mm
        if not self._ch.isOpen():
            return
        self._ch.sendCommand(CMD_ENABLE, AXIS_LATERAL, 1)
        self.rehome()

    def rehome(self):
        if not self._ch.isOpen():
            return
        self._state = HOMING
        self._ch.sendCommand(CMD_HOME_START, AXIS_LATERAL)

    # -- State machine tick (call every control loop) -------------------
    def update(self):
        if not self._ch.isOpen():
            return
        self._telem = self._ch.readLateralTelem()
        st     = self._telem.get('state', 0)
        faults = self._telem.get('faults', 0)

        if faults & FAULT_HOME_SENSOR:
            self._state = FAULT
            return
        if (faults & FAULT_OVERRUN) and self._state not in (HOMING, HOMING_OFFSET):
            self._state = FAULT
            return

        pos_mm  = self.get_current_position_mm()
        running = bool(st & STATE_RUNNING)
        at_home = bool(st & STATE_AT_HOME)

        if self._state == HOMING:
            # PRU1 drives backward and stops at sensor; host waits for idle
            if at_home and not running:
                self._state = HOMING_OFFSET
                if self._home_offset_mm > 0.001:
                    _push_move_mm(self._ch, LAT_HOME_SPEED_HZ // 2, True,
                                  self._home_offset_mm)
                else:
                    self._ch.sendCommand(CMD_RESET_POSITION, AXIS_LATERAL)
                    self._state = HOMED

        elif self._state == HOMING_OFFSET:
            if not running:
                self._ch.sendCommand(CMD_RESET_POSITION, AXIS_LATERAL)
                self._state = HOMED

        elif self._state == POSITIONING:
            if not running:
                self._state = HOMED

        elif self._state == WINDING_FWD:
            if pos_mm >= self._end_mm:
                self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL)
                if self._stop_on_next_high:
                    self._stop_on_next_high = False
                    self._state = HOMED
                elif self._pause_on_next_reversal:
                    self._pause_on_next_reversal = False
                    self._paused_at_reversal = True
                    self._state = HOMED
                else:
                    self._pass_count += 1
                    width = self._end_mm - self._start_mm
                    _push_move_mm(self._ch, self._last_lat_hz, False, width)
                    _push_constant_ms(self._ch, self._last_lat_hz, False, _REFILL_MS)
                    self._winding_fwd = False
                    self._state = WINDING_BWD

        elif self._state == WINDING_BWD:
            if pos_mm <= self._start_mm or at_home:
                self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL)
                if self._stop_on_next_low:
                    self._stop_on_next_low = False
                    self._state = HOMED
                else:
                    self._pass_count += 1
                    width = self._end_mm - self._start_mm
                    _push_move_mm(self._ch, self._last_lat_hz, True, width)
                    _push_constant_ms(self._ch, self._last_lat_hz, True, _REFILL_MS)
                    self._winding_fwd = True
                    self._state = WINDING_FWD

    # -- Getters --------------------------------------------------------
    def get_state(self) -> str: return self._state

    def is_homed(self) -> bool:
        return self._state in (HOMED, WINDING_FWD, WINDING_BWD)

    def is_fault(self) -> bool: return self._state == FAULT

    def is_busy(self) -> bool:
        return bool(self._telem.get('state', 0) & STATE_RUNNING)

    def is_traversing(self) -> bool:
        return self._state in (WINDING_FWD, WINDING_BWD)

    def is_at_zero(self) -> bool:
        return bool(self._telem.get('state', 0) & STATE_AT_HOME)

    def is_positioned_for_start(self) -> bool:
        return (self._state == HOMED and
                abs(self.get_current_position_mm() - self._start_pos_mm) < 0.1)

    def get_current_position_mm(self) -> float:
        return _steps_to_mm(self._telem.get('position_steps', 0))

    def get_target_position_mm(self) -> float:
        return self._start_pos_mm

    def get_traversal_progress(self) -> float:
        if not self.is_traversing() or abs(self._end_mm - self._start_mm) < 0.001:
            return 0.0
        pos = self.get_current_position_mm()
        return max(0.0, min(1.0, (pos - self._start_mm) / (self._end_mm - self._start_mm)))

    def get_pass_count(self) -> int: return self._pass_count

    def get_actual_velocity_hz(self) -> int:
        iv = self._telem.get('current_interval', 0)
        if not iv:
            return 0
        return PRU_CLOCK_HZ // (2 * iv)

    def get_instantaneous_lat_hz_nominal(self) -> int: return self._last_lat_hz
    def get_home_offset(self) -> float: return self._home_offset_mm
    def set_home_offset(self, mm: float): self._home_offset_mm = float(mm)

    def consume_paused_at_reversal(self) -> bool:
        v = self._paused_at_reversal
        if v:
            self._paused_at_reversal = False
        return v

    # -- One-shot flags -------------------------------------------------
    def arm_stop_at_next_high(self): self._stop_on_next_high = True
    def arm_stop_at_next_low(self):  self._stop_on_next_low  = True
    def arm_pause_on_next_reversal(self): self._pause_on_next_reversal = True
    def clear_one_shot_stops(self):
        self._stop_on_next_high      = False
        self._stop_on_next_low       = False
        self._pause_on_next_reversal = False
        self._paused_at_reversal     = False

    # -- Motion commands ------------------------------------------------
    def park_at_zero(self):
        self.prepare_start_position(0.0)

    def stop_winding(self):
        self.stop()

    def stop(self):
        if not self._ch.isOpen():
            return
        self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL)
        if self._state in (WINDING_FWD, WINDING_BWD):
            self._state = HOMED

    def prepare_start_position(self, target_mm: float, speed_hz: int = LAT_HOME_SPEED_HZ):
        if not self._ch.isOpen():
            return
        self._start_pos_mm = float(target_mm)
        current = self.get_current_position_mm()
        dist = target_mm - current
        if abs(dist) < 0.01:
            return
        self._state = POSITIONING
        fwd = dist > 0
        _push_move_mm(self._ch, speed_hz, fwd, abs(dist))

    def jog(self, delta_mm: float):
        if not self._ch.isOpen() or delta_mm == 0:
            return
        fwd = delta_mm > 0
        _push_move_mm(self._ch, LAT_HOME_SPEED_HZ // 4, fwd, abs(delta_mm))

    def start_winding(self, hz_nominal: int, tpp: float,
                      start_mm: float, end_mm: float, speed_scale: float = 1.0):
        if not self._ch.isOpen():
            return
        self._start_mm = float(start_mm)
        self._end_mm   = float(end_mm)
        lat_hz = self._calc_lat_hz(hz_nominal, tpp, end_mm - start_mm, speed_scale)
        self._last_lat_hz = lat_hz
        width = end_mm - start_mm
        _push_move_mm(self._ch, lat_hz, True, width)
        _push_constant_ms(self._ch, lat_hz, True, _REFILL_MS)
        self._winding_fwd = True
        self._state = WINDING_FWD

    def update_winding(self, hz_nominal: int, tpp: float,
                       start_mm: float, end_mm: float, speed_scale: float = 1.0):
        if not self._ch.isOpen():
            return
        if start_mm >= end_mm or tpp <= 0:
            return
        lat_hz = self._calc_lat_hz(hz_nominal, tpp, end_mm - start_mm, speed_scale)
        if lat_hz == self._last_lat_hz:
            return
        self._last_lat_hz = lat_hz
        # Flush existing moves and repush at new speed
        self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL)
        width = self._end_mm - self._start_mm
        _push_move_mm(self._ch, lat_hz, self._winding_fwd, width)
        _push_constant_ms(self._ch, lat_hz, self._winding_fwd, _REFILL_MS)

    # -- Private helpers ------------------------------------------------
    @staticmethod
    def _calc_lat_hz(spindle_hz: int, tpp: float, eff_width_mm: float,
                     speed_scale: float) -> int:
        """Lateral frequency = effWidth_mm * spindle_Hz / (tpp * STEPS_PER_REV) * scale."""
        if tpp <= 0 or SPINDLE_STEPS_PER_REV == 0 or eff_width_mm <= 0:
            return _LAT_MIN_HZ
        lat_hz = int(eff_width_mm * spindle_hz / (tpp * SPINDLE_STEPS_PER_REV) * speed_scale)
        return max(_LAT_MIN_HZ, lat_hz)

File: src/linux/src/LateralPRU.cpp
/* LateralPRU.cpp — Lateral carriage API delegating to PRU1 via IpcChannel.
 *
 * Klipper-style: all moves pre-computed as (interval, count, add) on host
 * and pushed to the PRU1 move ring.  No Hz/start/stop speed commands.
 * Homing: CMD_HOME_START tells PRU1 to enter homing_mode; PRU1 auto-stops
 * and resets position when the home sensor fires.
 */

#include "LateralPRU.h"
#include <cmath>
#include <cstring>

/* ── Hardware constants ──────────────────────────────────────────────────────*/
static constexpr uint32_t HOMING_STEPS = 700000u; /* ~228 mm > max travel   */

LateralPRU::LateralPRU(IpcChannel &channel) : _channel(channel) {}

/* ── Helpers ─────────────────────────────────────────────────────────────────*/
uint32_t LateralPRU::_mmToSteps(float mm) {
    return (uint32_t)(mm * (float)LAT_STEPS_PER_MM + 0.5f);
}

float LateralPRU::_stepsToMm(int32_t steps) {
    return (float)steps / (float)LAT_STEPS_PER_MM;
}

bool LateralPRU::_pushTraverse(float fromMm, float toMm, uint32_t latHz) {
    float delta = toMm - fromMm;
    if (latHz == 0u || std::abs(delta) < 0.001f) return false;
    bool fwd   = (delta > 0.0f);
    float absd = fwd ? delta : -delta;
    uint32_t n_edges = 2u * _mmToSteps(absd);
    if (n_edges == 0u) return false;
    return MoveQueue::pushConstantEdges(_channel, AXIS_LATERAL, latHz, fwd, n_edges);
}

/* ── State machine ───────────────────────────────────────────────────────────*/
void LateralPRU::begin(float homeOffsetMm) {
    _homeOffsetMm = homeOffsetMm;
    if (!_channel.isOpen()) return;
    _channel.sendCommand(CMD_ENABLE, AXIS_LATERAL, 1u);
    rehome();
}

void LateralPRU::rehome() {
    if (!_channel.isOpen()) return;
    _state = LatState::HOMING;
    _channel.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL);
    _channel.sendCommand(CMD_HOME_START,  AXIS_LATERAL);
    /* Push many slow backward edges; PRU1 flushes ring on HOME_HIT */
    MoveQueue::pushConstantEdges(_channel, AXIS_LATERAL,
                                  LAT_HOME_HZ, false, 2u * HOMING_STEPS);
}

void LateralPRU::update() {
    if (!_channel.isOpen()) return;
    _refreshTelem();

    uint16_t st     = _telem.state;
    uint16_t faults = _telem.faults;

    if (faults & FAULT_HOME_SENSOR) {
        _state = LatState::FAULT; return;
    }
    if ((faults & FAULT_OVERRUN) &&
        _state != LatState::HOMING && _state != LatState::HOMING_OFFSET) {
        _state = LatState::FAULT; return;
    }

    bool motor_idle = !(st & STATE_RUNNING) && (_telem.moves_pending == 0u);

    switch (_state) {

    case LatState::HOMING:
        /* PRU1 stops and resets position when HOME_HIT. */
        if ((st & STATE_AT_HOME) && motor_idle) {
            _state = LatState::HOMING_OFFSET;
            _pushTraverse(0.0f, _homeOffsetMm, LAT_HOME_HZ / 2u);
        }
        break;

    case LatState::HOMING_OFFSET:
        if (motor_idle) {
            _channel.sendCommand(CMD_RESET_POSITION, AXIS_LATERAL);
            _state = LatState::HOMED;
        }
        break;

    case LatState::POSITIONING:
        if (motor_idle)
            _state = LatState::HOMED;
        break;

    case LatState::WINDING_FWD:
        if (motor_idle) {
            if (_stopOnNextHigh) {
                _stopOnNextHigh = false;
                _state = LatState::HOMED;
            } else if (_pauseOnNextReversal) {
                _pauseOnNextReversal = false;
                _pausedAtReversal    = true;
                _state = LatState::HOMED;
            } else {
                /* Push full backward traverse */
                _pushTraverse(_endMm, _startMm, _windingLatHz);
                _state = LatState::WINDING_BWD;
            }
        }
        break;

    case LatState::WINDING_BWD:
        if (motor_idle || (st & STATE_AT_HOME)) {
            if (_stopOnNextLow) {
                _stopOnNextLow = false;
                _state = LatState::HOMED;
            } else {
                /* Push full forward traverse */
                _pushTraverse(_startMm, _endMm, _windingLatHz);
                _state = LatState::WINDING_FWD;
            }
        }
        break;

    default:
        break;
    }
}

void LateralPRU::prepareStartPosition(float startMm, uint32_t speedHz) {
    if (!_channel.isOpen()) return;
    _startPosMm = startMm;
    _refreshTelem();
    float current = _stepsToMm(_telem.position_steps);
    if (std::abs(startMm - current) < 0.05f) return;
    _state = LatState::POSITIONING;
    _channel.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL);
    _pushTraverse(current, startMm, speedHz > 0u ? speedHz : LAT_HOME_HZ);
}

void LateralPRU::jog(float deltaMm) {
    if (!_channel.isOpen() || std::abs(deltaMm) < 0.001f) return;
    _refreshTelem();
    float current = _stepsToMm(_telem.position_steps);
    _pushTraverse(current, current + deltaMm, LAT_HOME_HZ / 4u);
}

void LateralPRU::startWinding(uint32_t windingHz, long tpp,
                               float startMm, float endMm, float speedScale) {
    if (!_channel.isOpen()) return;
    _startMm = startMm;
    _endMm   = endMm;

    float effWidth = endMm - startMm;
    if (effWidth <= 0.0f || tpp <= 0) return;

    /* Lateral Hz = effWidthMm × windingHz / (tpp × STEPS_PER_REV) */
    _windingLatHz = (uint32_t)(effWidth * (float)windingHz /
                               ((float)tpp * (float)LAT_STEPS_PER_MM) * speedScale);
    if (_windingLatHz < 10u) _windingLatHz = 10u;

    _channel.sendCommand(CMD_QUEUE_FLUSH, AXIS_LATERAL);
    _pushTraverse(startMm, endMm, _windingLatHz);
    _state = LatState::WINDING_FWD;
}

void LateralPRU::updateWinding(uint32_t windingHz, long tpp, float speedScale) {
    if (_startMm >= _endMm || tpp <= 0) return;
    float effWidth = _endMm - _startMm;
    uint32_t latHz = (uint32_t)(effWidth * (float)windingHz /
                                ((float)tpp * (float)LAT_STEPS_PER_MM) * speedScale);
    if (latHz < 10u) latHz = 10u;
    /* Speed change will apply on the next traversal push (at boundary) */
    _windingLatHz = latHz;
}

/* ── State queries ───────────────────────────────────────────────────────────*/
bool LateralPRU::isHomed() const {
    return _state == LatState::HOMED
        || _state == LatState::WINDING_FWD
        || _state == LatState::WINDING_BWD;
}

bool LateralPRU::isBusy() const {
    _refreshTelem();
    return (_telem.state & STATE_RUNNING) != 0u;
}

bool LateralPRU::isAtZero() const {
    return (_telem.state & STATE_AT_HOME) != 0u;
}

bool LateralPRU::isPositionedForStart() const {
    return _state == LatState::HOMED
        && std::abs(getPositionMm() - _startPosMm) < 0.1f;
}

float LateralPRU::getPositionMm() const {
    _refreshTelem();
    return _stepsToMm(_telem.position_steps);
}

uint32_t LateralPRU::getCurrentLateralHz() const {
    _refreshTelem();
    uint32_t iv = _telem.current_interval;
    if (iv == 0u) return 0u;
    return PRU_CLOCK_HZ / (2u * iv);
}

/* ── One-shot stop flags ─────────────────────────────────────────────────────*/
void LateralPRU::clearOneShotStops() {
    _stopOnNextHigh = _stopOnNextLow = _pauseOnNextReversal = _pausedAtReversal = false;
}

bool LateralPRU::hasStopAtNextBoundArmed() const {
    return _stopOnNextHigh || _stopOnNextLow || _pauseOnNextReversal;
}

bool LateralPRU::consumePausedAtReversal() {
    bool v = _pausedAtReversal; _pausedAtReversal = false; return v;
}

/* ── stateStr ────────────────────────────────────────────────────────────────*/
const char *LateralPRU::stateStr() const {
    switch (_state) {
    case LatState::FAULT:          return "FAULT";
    case LatState::HOMING:         return "HOMING";
    case LatState::HOMING_OFFSET:  return "HOMING_OFFSET";
    case LatState::HOMED:          return "HOMED";
    case LatState::POSITIONING:    return "POSITIONING";
    case LatState::WINDING_FWD:    return "WINDING_FWD";
    case LatState::WINDING_BWD:    return "WINDING_BWD";
    default:                       return "UNKNOWN";
    }
}

void LateralPRU::_refreshTelem() const {
    _channel.readLateralTelem(_telem);
}




